src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java

```diff
@@ -0,0 +1,113 @@
+import com.blamejared.crafttweaker.impl.brackets.RecipeTypeBracketHandler;
+import com.blamejared.crafttweaker.impl.recipes.wrappers.WrapperRecipe;
+import com.blamejared.crafttweaker_annotations.annotations.Document;
+import com.mojang.datafixers.util.Pair;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.util.registry.Registry;
+import org.openzen.zencode.java.ZenCodeType;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@ZenRegister
+@ZenCodeType.Name("crafttweaker.api.recipes.Replacer")
+@Document("vanilla/api/recipes/Replacer")
+public final class Replacer {
+   
+    private final Collection<IRecipeManager> targetedManagers;
+    private final Collection<? extends IRecipe<?>> targetedRecipes;
+    private final List<ReplacementRule> rules;
+    
+    private Replacer(final Collection<? extends IRecipe<?>> recipes, final Collection<IRecipeManager> managers) {
+        this.targetedManagers = Collections.unmodifiableCollection(managers);
+        this.targetedRecipes = Collections.unmodifiableCollection(recipes);
+        this.rules = new ArrayList<>();
+    }
+    
+    @ZenCodeType.Method
+    public static Replacer forRecipes(final WrapperRecipe... recipes) {
+        if (recipes.length <= 0) {
+            throw new IllegalArgumentException("Unable to create a replacer without any targeted recipes");
+        }
+        return new Replacer(Arrays.stream(recipes).map(WrapperRecipe::getRecipe).collect(Collectors.toList()), Collections.emptyList());
+    }
+    
+    @ZenCodeType.Method
+    public static Replacer forTypes(final IRecipeManager... managers) {
+        if (managers.length <= 0) {
+            throw new IllegalArgumentException("Unable to create a replacer without any targeted recipe types");
+        }
+        return new Replacer(Collections.emptyList(), Arrays.asList(managers));
+    }
+    
+    @ZenCodeType.Method
+    public static Replacer forAllTypes() {
+        return new Replacer(Collections.emptyList(), RecipeTypeBracketHandler.getManagerInstances());
+    }
+    
+    @ZenCodeType.Method
+    public Replacer replace(final IIngredient from, final IIngredient to) {
+        // TODO("Avoid trivial replacement rules -- check if from == to")
+        if (from.contains(to) && to.contains(from)) return this;
+        this.rules.add(new ReplacementRule(from, to));
+        return this;
+    }
+    
+    @ZenCodeType.Method
+    public void execute() {
+        if (this.rules.isEmpty()) return;
+        
+        final List<ReplacementRule> rules = Collections.unmodifiableList(this.rules);
+
+        Stream.concat(this.streamManagers(), this.streamRecipes())
+                .map(pair -> this.execute(pair.getFirst(), pair.getSecond(), rules))
```

Comment:
In regards to the questions earlier (duplicates)
.distinct before map?

Reply:
Mangers and recipes are now stored in a `Set`, so this should guarantee the lack of duplicates